### ISSUE TITLE
DMD tests fail on some targets

### ISSUE THREAD

**Issue by leo60228** [0x0ade/MonoMod#40]
From log of [build 216](https://dev.azure.com/0x0ade/MonoMod/_build/results?buildId=216):
```
2019-02-07T14:38:53.9276265Z Results File: D:\a\_temp\VssAdministrator_fv-az576_2019-02-07_14_38_53.trx
2019-02-07T14:38:53.9284439Z 
2019-02-07T14:38:53.9286042Z Total tests: 7. Passed: 7. Failed: 0. Skipped: 0.
2019-02-07T14:38:53.9286274Z Test Run Successful.
2019-02-07T14:38:53.9286489Z Test execution time: 8.2340 Seconds
2019-02-07T14:38:53.9977187Z ##[error]Error: C:\hostedtoolcache\windows\dncs\2.0.0\x64\dotnet.exe failed with return code: 1
2019-02-07T14:38:55.6017726Z ##[error]Dotnet command failed with non-zero exit code on the following projects : D:\a\1\s\MonoMod.UnitTest\MonoMod.UnitTest.csproj
```

**Comment by leo60228** [comment]
Closing, I misread the logs. Sorry.

**Comment by 0x0ade** [comment]
There's still an issue though, either regarding the test environment, or a general regression.

I'll keep this issue open until all test issues are resolved.

**Comment by leo60228** [comment]
Fixed by 89a6454cf19556d90d3c6c90c7e56474872c3324.

**Comment by 0x0ade** [comment]
I've now fixed all async-related DynamicMethodDefinition issues by locking the module access more aggressively and cloning the underlying MethodDefinition on access.

This effectively changes the behavior, as multiple DynamicMethodDefinitions no longer modify the same underlying MethodDefinition. I still don't see it as a API-breaking change though, as we're trading shared memory for proper reloading and separate patches.

Build result: [https://0x0ade.visualstudio.com/MonoMod/_build/results?buildId=219](https://0x0ade.visualstudio.com/MonoMod/_build/results?buildId=219)
